components/authz-service/server/v2/policy_refresher.go

```diff
@@ -69,7 +75,18 @@ RUNLOOP:
 		case <-ctx.Done():
 			refresher.log.WithError(ctx.Err()).Info("Policy refresher exiting")
 			break RUNLOOP
+		case <-refresher.changeNotifier.C():
+			refresher.log.Info("Received policy change notification")
+			var err error
+			lastPolicyID, err = refresher.refresh(context.Background(), lastPolicyID)
+			if err != nil {
+				refresher.log.WithError(err).Warn("Failed to refresh policies")
+			}
+			if !antiEntropyTimer.Stop() {
+				<-antiEntropyTimer.C
```

Comment:
how is this timer used? it seems like the refresher is both listening for change notifications and also periodically refreshing based on the timer - is that right? are both needed?

Reply:
the reason for both is you have to assume that notifications can be missed for various reasons. For example, you drop the connection and a notification comes in, you will miss it. Doing this is much easier than trying to guarantee delivery notifications and it's pretty cheap to do. The notifications just become an optimization that speeds up replication.